metals/src/main/scala/scala/meta/internal/worksheets/MetalsWorksheetProvider.scala

```diff
@@ -0,0 +1,352 @@
+import scala.collection.concurrent.TrieMap
+import ch.epfl.scala.bsp4j.BuildTargetIdentifier
+import mdoc.internal.cli.Context
+import mdoc.internal.markdown.SectionInput
+import mdoc.internal.markdown.Modifier
+import mdoc.internal.markdown.Instrumenter
+import mdoc.internal.markdown.MarkdownCompiler
+import org.eclipse.{lsp4j => l}
+import scala.meta.internal.decorations.ThemableDecorationInstanceRenderOptions
+import scala.meta.internal.decorations.ThemableDecorationAttachmentRenderOptions
+import mdoc.internal.cli.Settings
+import scala.meta.internal.metals.ScalaTarget
+import scala.concurrent.Future
+import scala.meta.pc.CancelToken
+import pprint.TPrintColors
+import scala.meta.internal.metals.UserConfiguration
+import scala.collection.immutable.Nil
+import mdoc.document.Statement
+import scala.meta.internal.metals.MetalsLanguageClient
+import scala.meta.internal.metals.ScalaVersions
+import scala.meta.internal.metals.BuildInfo
+import org.eclipse.lsp4j.PublishDiagnosticsParams
+import scala.meta.internal.pc.CompilerJobQueue
+import java.util.concurrent.CompletableFuture
+import scala.concurrent.ExecutionContext
+import java.util.concurrent.ScheduledExecutorService
+import scala.meta.internal.metals.MetalsSlowTaskParams
+import java.util.concurrent.TimeUnit
+import scala.meta.internal.metals.MutableCancelable
+import scala.meta.internal.metals.StatusBar
+import scala.meta.internal.pc.InterruptException
+import scala.util.control.NonFatal
+
+/**
+ * Implements interactive worksheets for "*.worksheet.sc" file extensions.
+ *
+ * Code is evaluated on file save using mdoc: https://scalameta.org/mdoc/
+ */
+class MetalsWorksheetProvider(
+    workspace: AbsolutePath,
+    buffers: Buffers,
+    buildTargets: BuildTargets,
+    languageClient: MetalsLanguageClient,
+    userConfig: () => UserConfiguration,
+    sh: ScheduledExecutorService,
+    statusBar: StatusBar
+)(implicit ec: ExecutionContext)
+    extends WorksheetProvider
+    with Cancelable {
+  // Worksheet evaluation happens on a single threaded job queue. Jobs are
+  // prioritized using the same order as completion/hover requests:
+  // first-come last-out.
+  private val jobs = CompilerJobQueue()
+  private val cancelables = new MutableCancelable()
+  private val contexts = new TrieMap[BuildTargetIdentifier, Context]()
+  private val reporter = new StoreReporter()
+
+  def clear(): Unit = {
+    jobs.shutdown()
+    contexts.values.foreach { context =>
+      context.compiler.global.close()
+    }
+    contexts.clear()
+  }
+  def cancel(): Unit = {
+    clear()
+  }
+
+  override def decorations(
+      path: AbsolutePath,
+      token: CancelToken
+  ): Future[Array[DecorationOptions]] = {
+    if (path.extension != "sc") {
+      Future.successful(Array.empty)
+    } else {
+      reporter.reset()
+      val result = new CompletableFuture[Array[DecorationOptions]]()
+      token.onCancel().asScala.foreach {
+        case java.lang.Boolean.TRUE =>
+          // Don't publish any
```

Comment:
the comment seems 'interrupted', was it meant to be

```suggestion
          // Don't publish any result on cancelation
```

Reply:
Fixed.